src/librustc/traits/specialize/specialization_graph.rs

```diff
@@ -291,8 +335,52 @@ impl<'a, 'gcx, 'tcx> Graph {
 
     /// The parent of a given impl, which is the def id of the trait when the
     /// impl is a "specialization root".
-    pub fn parent(&self, child: DefId) -> DefId {
-        *self.parent.get(&child).unwrap()
+    pub fn parent(&self, child: DefId) -> Vec<DefId> {
+        self.parent.get(&child).unwrap().clone()
+    }
+
+    pub fn build_graph(&self) -> (DataStructuresGraph<DefId, String>, DefIdMap<NodeIndex>) {
+        let mut sg_graph: DataStructuresGraph<DefId, String> = DataStructuresGraph::new();
+        let mut nodes_idx = Default::default();
+        for (key, val) in self.parent.iter() {
+            let idx = self.node_idx(&mut sg_graph, &mut nodes_idx, *key);
+            for parent in val {
+                let pidx = self.node_idx(&mut sg_graph, &mut nodes_idx, *parent);
+                sg_graph.add_edge(idx, pidx, format!("fromt {:?} to {:?}", key, parent));
+            }
+        }
+
+        (sg_graph, nodes_idx)
+    }
+
+    /// Return true if impl1 and impl2 are allowed to overlap:
+    /// They have an intersection impl
+    pub fn check_overlap(&self,
+                         sg_graph: &DataStructuresGraph<DefId, String>,
+                         nodes_idx: &DefIdMap<NodeIndex>,
+                         impl1: DefId,
+                         impl2: DefId) -> bool {
+
+        let impl1_idx = *nodes_idx.get(&impl1).unwrap();
+        let impl2_idx = *nodes_idx.get(&impl2).unwrap();
+
+        let impl1_incoming_nodes = sg_graph.nodes_in_postorder(INCOMING, impl1_idx);
+        let impl2_incoming_nodes = sg_graph.nodes_in_postorder(INCOMING, impl2_idx);
+
+        impl1_incoming_nodes[0] == impl2_incoming_nodes[0]
+    }
+
+    fn node_idx(&self,
+                sg_graph: &mut DataStructuresGraph<DefId, String>,
+                nodes_idx: &mut DefIdMap<NodeIndex>,
+                node: DefId) -> NodeIndex {
+        if nodes_idx.get(&node).is_some() {
+            *nodes_idx.get(&node).unwrap()
+        } else {
+            let idx = sg_graph.add_node(node);
+            nodes_idx.insert(node, idx);
+            idx
+        }
```

Comment:
it feels like it'd be nice to extract out this graph into something we can unit-test, that is more independent of the compiler -- eg., living in `rustc_data_structures`